src/server/external-services/pathway-commons.js

```diff
@@ -121,23 +125,70 @@ const sifGraph = async ( queryObj ) => {
   });
 };
 
-/* xref2Uri
- * Light wrapper around pc metadata service for mapping Xrefs db, id to url
+const handleEntityUriResponse = text => {
+  if( !text ) return ''; // Go ahead so we can cache, but gotta throw to inform clients
+
+  const uri = new url.URL( text ); // Throws TypeError
+  const pathParts = _.compact( uri.pathname.split('/') );
+  if( _.isEmpty( pathParts ) || pathParts.length !== 2 ) throw new Error( 'Unrecognized URI' );
+  const namespace = _.head( pathParts );
+  return uri.origin + '/' + namespace;
+};
+
+const constructQueryPath = ( name, localId ) => {
+  // Edge case - localId has periods e.g. 'enzyme nomenclature/6.1.1.5' gotta add a trailing slash
+  const suffix = /\./.test( localId ) ? '/' : '';
+  return name + '/' + localId + suffix;
+};
+
+/* fetchEntityUriBase
+ * Light wrapper around the pc2 service to get the uri given a collection name and local ID for entity
  * http://www.pathwaycommons.org/pc2/swagger-ui.html#!/metadata45controller/identifierOrgUriUsingGET
+ * NB: pc2 service returns 200 and empty body if collection name and/or local ID are unrecognized.
+ *   If the local ID is empty, throws a 404
+ * @return { object } a node URL object
+ */
+const fetchEntityUriBase = ( name, localId ) => {
+  //console.log( `fetchEntityUriBase with collection = ${name}; ID = ${localId}` );
+  const url = config.PC_URL + 'pc2/miriam/uri/' + constructQueryPath( name, localId ) ;
+  return fetch( url , { method: 'GET', headers: { 'Accept': 'text/plain' } })
+    .then( res => res.text() )
+    .then( handleEntityUriResponse );
+};
+
+// Custom cache since 1. want to hash off only one param 2. cases I don't wish to delete on error
+// I hacked the cache previously to hash off a single param index, but wasn't sure...
+const rawGetEntityUriBase = ( name, localId ) => {
+  let hash = hasher.hash( name );
+  if( pcCache.has( hash ) ){
+    return pcCache.get( hash );
+  } else {
+    let res = fetchEntityUriBase( name, localId );
+    pcCache.set( hash, res );
+    res.catch( err => {
+      pcCache.del( hash );
+      logger.error(`Failed to fill cache with ${name} and ${localId} - ${err}`);
+    });
```

Comment:
If we don't already have a util function to cache a fetch call, then we should add one later to avoid repetition.  Factoid, for example, uses a util memoize function that accepts a custom cache per `memoize()` call.  Then you just wrap the fetch function.